### DISCUSSION TITLE
Allow importing own CA certificates prior to starting containers

### DISCUSSION THREAD

**Discussion by LuFlo** [nextcloud/all-in-one#805] [Ideas] [answered]
Basically the whole issue is described here: https://github.com/nextcloud/all-in-one/discussions/717

In short: It should be "officially" possible to use own certificates via a reverse proxy for the apache container by installing a CA file inside the mastercontainer. A bonus would be, if one could configure ssl certificates without using a reverse proxy, but this would mean more work I guess.

In the discussion mentioned above, I also posted a workaround. This could be a starting point for implementing this feature. I imagine that you could copy the CA file in a designated folder in the config volume of `nextcloud-aio-mastercontainer`. An amazing feature would be to copy the certificated via the web interface but let's make small steps 😅

What do you think about it?

**Comment by szaimen** [comment]
I will not work on this but PRs are welcome!

**Comment by dantefromhell** [comment]
Would be very helpful as it's fairly easy with some testing to blow through the rate limiting of ZeroSSL/ Let's Encrypt.

**Comment by szaimen** [comment]
What about an option for skipping the domain validation process. Would that cover your suitcase?

**Reply by LuFlo** [reply]
I actually searched the part which takes care of the domain validation in order to try this out first, but I didn't find it. Do you have the location of this particular code at hand?

**Reply by szaimen** [reply]
So basically these lines could be skipped then:
https://github.com/nextcloud/all-in-one/blob/2d9a1afa336e12686f15c451d6fdfa50dc2c0855/php/src/Data/ConfigurationManager.php#L201-L239

**Reply by szaimen** [reply]
Or maybe would mounting `/usr/share/ca-certificates` from the host into the same place inside the mastercontainer make it work already?
I could document this then...

**Reply by dantefromhell** [reply]
From what I could gather so far, certs get auto-requested by Caddy twice
- one in the `nextcloud-aio-mastercontainer` container if you connect to port `:8443`
- one in the `nextcloud-aio-apachage` container

correct?

So when trying to use a custom certificate (possibly incl. chain) + private key those would need to be inserted into both containers before startup, no?

**Reply by LuFlo** [reply]
> Or maybe would mounting `/usr/share/ca-certificates` from the host into the same place inside the mastercontainer make it work already? I could document this then...

Sorry for the late reply. Well there is another issue: I realized that we need to copy the CA into the talk container as well, otherwise the talk container does not accept the nextcloud container. Interestingly there is a checkbox in the Talk settings UI do disable SSL validation, but I think this only disables the validation of the signaling server but not the other way around. The thing is, the talk container does not seem to have a volume, so I had to paste the CA into a file manually. I guess it has to be done after every update though...

**Reply by szaimen** [reply]
So this is one of the reasons why I still think that allowing self-signed certificates is a bad idea since it comes with many unexpected side effects...

**Comment by szaimen** [comment]
This could be an alternative for self-signed certificates: https://github.com/nextcloud/all-in-one/issues/878

**Comment by szaimen** [comment]
I am honest to you: I do not want to support self-signed certificates so this will not be added. 

There are enough possibilities to get a valid certificate for AIO and there is now also a way to disable the domain validation.

**Comment by szaimen** [comment, marked as answer]
It is since a while possible to simply skip the domain validation which should resolve this.

**Reply by szaimen** [reply]
And it is even possible to mount a ca certificate into the Nextcloud container. E.g. for LDAPS
